### ISSUE TITLE
Support for action of block

### ISSUE THREAD

**Issue by catull** [schibsted/jslt#71]
I have a requirement to break up a text into lines of up to N characters.
These lines must contain full words.

E.g. if a line exceeds N characters, then the last word must be shifted to the next line.
Repeat this process until at most N characters are collected for the current line.

This is the input:
```json
{
  "lfn": "1169480",
  "publicationText": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur."
}
```

The non-working JSTL only captures the first line:
```
let words = split (.publicationText, " ")
let maxLineLength = 65

def lineBreak (nr, ws, rest)
  let line = trim (join ($ws, " "))
  let line2 = trim ($line + " " + $rest[0])
  // let len = size ($line)
  let len2 = size ($line2)

  if ($len2 > maxLineLength or not ($rest)) (
    { "SORLFN": .lfn, "SORZLN": $nr, "SORTXT": $line }
    // lineBreak ($nr + 1, [], $rest) // <--- this line is not accepted HERE
  ) else //if ($len2 < maxLineLength)
    lineBreak ($nr, $ws + $rest[0 : 1], $rest[1 : ])

{
  "ORIGINAL": lineBreak (0, [], $words)
}
```

Unfortunately, the second line in the positive `if`-block is not allowed.
Apparently `if` and `else` only allow a single line of expression.

I would need to have 2 expressions in the `if`-block.

Is there an alternative ?

**Comment by catull** [comment]
Found a nice solution:

```
let elephant = .lfn
let words = split (.publicationText, " ")
let maxLineLength = 65

def lineBreak (result, nr, ws, rest)
  let line  = join ($ws, " ")
  let lineR = join ($ws + $rest, " ")
  let line2 = $line + " " + $rest[0]
  // let len   = size ($line)
  let len2  = size ($line2)
  let lenr  = size ($lineR)
  // let rsiz  = size ($ws)
  let lineO = if ($lenr != $len2) $line else $lineR

  let obj = {
    "SORLFN": $elephant,
    "SORZLN": $nr,
    "SORTXT": $lineO
    // , "line2": $line2, "ws": $ws, "lineR": $lineR, "len": $len, "len2": $len2, "rsiz": $rsiz, "lenr": $lenr
  }

  if (not ($rest))
    $result
  else if ($lenr == $len2)
    lineBreak ($result + [ $obj ], 0, [], [])
  else if ($len2 > $maxLineLength)
    lineBreak ($result + [ $obj ], $nr + 1, $rest[0 : 1], $rest[1 : ])
  else
    lineBreak ($result, $nr, $ws + $rest[0 : 1], $rest[1 : ])

{
  "ORIGINAL": lineBreak ([], 0, [], $words)
}
```

**Comment by larsga** [comment]
The problem with

```
    if ($len2 > maxLineLength or not ($rest)) (
        { "SORLFN": .lfn, "SORZLN": $nr, "SORTXT": $line }
        // lineBreak ($nr + 1, [], $rest) // <--- this line is not accepted HERE
  ) else ...
```

is that it's not clear what it means. In general `if (<condition>) <expr> else <expr>` produces the value of one of the two expressions. If you have two expressions in the same branch, one of the two is pointless, because the `if` can only produce a single value.

Anyway, I'm glad you found a solution. I suppose that means this issue can be closed?

**Comment by catull** [comment]
You do see the irony here ?

I asked for support of block actions.

You do point out
> [ .... ] because the if can only produce a single value.

I can confirm this is the case.

Block actions means that a combinations of producing JSON "lines", as well as other statements, such as builtin or locally defined function.

Although I found a solution, the original support request was not addressed.

I suggest you explain why or why not "block actions" - i.e. more than one statement - are supported or not.

In other words, you are free to close this issue stating that block actions - or multiple statements in `if` branches (both positive and negative) are going to be implemented/planned
- never
- not in the near future
- or not before version x.y

**Comment by larsga** [comment]
My problem is I don't understand what you mean by "block actions", so right now I have no idea whether it belongs in the language or not.

I don't understand this at all:

> Block actions means that a combinations of producing JSON "lines", as well as other statements, such as builtin or locally defined function.

To me it looks like instead of writing one expression you write two expressions. But since all of JSLT is based on expressions producing single values it's not clear to me what this would mean. The definition of "value" here is a JSON value (something that can be serialized to JSON).

To take an example, if you write

```
if (true) (
  1
  2
)
```

what is the output? It has to be 1 or 2. It can't be both.

To be clearer: if you could return 1 *and* 2 at the same time, JSLT expressions operating on this run into trouble. How do you operate on two values simultaneously? What does

```
3 + if (true) (
  1
  2
)
```

mean?

**Comment by larsga** [comment]
I think it's not necessarily the best thing to focus on the "block action" concept directly. It may be more helpful if you can explain what you want to use block actions for. What problem do they solve? I find it's often easier to understand what people want by approaching it this way.

**Comment by catull** [comment]
Apologies for not being clear.

Originally in this code snippet

>     if ($len2 > maxLineLength or not ($rest)) (
>        { "SORLFN": .lfn, "SORZLN": $nr, "SORTXT": $line }
>        // lineBreak ($nr + 1, [], $rest) // <--- this line is not accepted HERE
>  ) else ...

The line collected so far contains less than 65 characters - including space characters.
It is represented by the `$line` variable.
`$line2` is `$line` with the next word appended; the length of `$line2` exceeds `maxLineLength` (65 characters.

At this point, I wanted to produce the JSON object like this one:
```json
{ "SORLFN": 132352347,
    "SORZLN": 0,
    "SORTXT": "First 65 character line"
}
```
At the same time, I wanted to recursively call the function again, increasing the line number, which will be captured in the next JSON object in the attribute `SORZLN`.

Perhaps it would help if the JSON object production could be expressed this way, although the grammar would change in a big way:

```
if (condition) (
  createJsonObject ({  "SORLFN": 132352347, "SORZLN": 0, "SORTXT": "First 65 character line" }) // STATEMENT 1
  someOtherFunction ($nr + 1, [], $rest) // STATEMENT 2
  // possibly other function calls
) else (
 // other statement(s)
)
```

This would require a change in the grammar.

Although I did find an alternative, there may not always be one; basically I combined the two statements above (STATEMENT 1 and 2) into one.

For these cases where an alternative is not possible, multiple function calls in `if` or `else` branch would be a great help.

I called the "multiple function calls" _block actions_, because I considered the single function in `if` or `else` an action in a block.

**Comment by larsga** [comment]
No worries about having difficulties explaining. This stuff is hard.

I'm still confused, because I don't understand what you're returning. So this code

```
f (condition) (
  createJsonObject ({  "SORLFN": 132352347, "SORZLN": 0, "SORTXT": "First 65 character line" }) // STATEMENT 1
  someOtherFunction ($nr + 1, [], $rest) // STATEMENT 2
  // possibly other function calls
)
```

how would you use the result? Let's say this is part of the definition of the function `foo`, and I do

```
let v = foo(<params>)
```

After this, how do I make use of `$v` and get access to the results of `createJsonObject` and `someOtherFunction`?
